system/protocols/generic/factory.py

```diff
@@ -0,0 +1,225 @@
+
+    # Base Twisted functions - override if necessary
+
+    def buildProtocol(self, addr):
+        """
+        Build a protocol instance for Twisted to use.
+
+        If your protocol uses the Ultros default *__init__* function, you will
+        likely not need to override this. For reference, the arguments expected
+        for that function are **(*name*, *factory*, *config*)**.
+        """
+
+        self.load_module()
+        self.protocol = self.protocol_module.Protocol(
+            self.name, self, self.config
+        )
+        return self.protocol
+
+    # Base Twisted functions - override if necessary and call superclass
+
+    def clientConnected(self):
+        """
+        Called when the protocol makes a successful connection.
+
+        If for some reason you aren't using the reactor for your connection,
+        then you may want to call this in your protocol manually, as it will
+        reset the reconnection counter as appropriate.
+        """
+
+        if self.reconnection_config["reset-on-success"]:
+            self.reconnection_attempts = 0
+
+    def clientConnectionFailed(self, connector, reason):
+        """
+        Called when the protocol failed to connect.
+
+        If for some reason you aren't using the reactor for your connection,
+        then you may want to call this in your protocol manually, as it
+        automatically handles reconnections as appropriate.
+        """
+
+        if self.shutting_down:
+            return
+
+        self.logger.warn("Connection failed: {}".format(reason))
+
+        if self.reconnection_config["on-failure"]:
+            if (
+                    self.reconnection_attempts >=
+                    self.reconnection_config["attempts"]
+            ):
+                self.logger.warn(
+                    "Unable to connect after {} attempts, aborting".format(
+                        self.reconnection_attempts
+                    )
+                )
+            return
+
+        self.reconnection_attempts += 1
+        delay = self.reconnection_config["delay"]
+        delay *= self.reconnection_attempts
+
+        self.logger.info(
+            "Connecting after {} seconds (attempt {}/{}".format(
+                delay, self.reconnection_attempts,
+                self.reconnection_config["attempts"]
+            )
+        )
+
+        reactor.callLater(delay, connector.connect)
+
+    def clientConnectionLost(self, connector, reason):
+        """
+        Called when the protocol connected, and then lost connection later.
+
+        If for some reason you aren't using the reactor for your connection,
+        then you may want to call this in your protocol manually, as it
+        automatically handles reconnections as appropriate.
+        """
+
+        if self.shutting_down:
+            return
+
+        self.logger.warn("Connection lost: {}".format(reason))
+
+        if self.reconnection_config["on-drop"]:
```

Comment:
Why would you not want to reconnect if it fails? It could be a brief network outage (perhaps the same thing that caused the initial drop), or perhaps the server has gone down for maintenance.

Ah, there's an indentation difference in the two different reconnect blocks that you may want to look at. In `clientConnectionFailed`, the if block doesn't include the reconnect logic, but it does in `clientConnectionLost`. I'll clarify what I meant once I know which of these two indentations is correct.